tests/catch/unit/stream/hipStreamCreateWithFlags.cc

```diff
@@ -17,23 +17,146 @@ OUT OF OR INN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+  using clock = std::chrono::high_resolution_clock;
+  const auto end = clock::now() + timeout;
+  bool finished = hipStreamQuery(stream) == hipSuccess;
+  while (!finished && clock::now() < end) {
+    std::this_thread::sleep_for(queryPeriod);
+    finished = hipStreamQuery(stream) == hipSuccess;
+  }
+  return !finished;
+}
+
+void checkConcurrentSupported() {
+  int device{};
+  hipGetDevice(&device);
+  hipDeviceProp_t deviceProps{};
+  hipGetDeviceProperties(&deviceProps, device);
+  REQUIRE(deviceProps.concurrentKernels == 1);
+}
+
+
+TEST_CASE("Unit_hipStreamCreateWithFlags_BlockNullStream") {
+  checkConcurrentSupported();
+
+  constexpr auto data_size = sizeof(int);
+  hipStream_t defaultStream = GENERATE(static_cast<hipStream_t>(nullptr), hipStreamPerThread);
+  INFO("Testing with default stream " << defaultStream);
+
+  int* device_data{};
+  HIP_CHECK(hipMalloc(&device_data, data_size));
+  HIP_CHECK(hipMemset(device_data, 0, data_size));
+
+  hipStream_t comparingStream{};
+  HIP_CHECK(hipStreamCreateWithFlags(&comparingStream, hipStreamDefault));
+  hipStream_t settingStream{};
+  HIP_CHECK(hipStreamCreateWithFlags(&settingStream, hipStreamNonBlocking));
+
+  // kernel will wait until the setting kernel sets some memory
+  hipLaunchKernelGGL(comparing_kernel, dim3(1), dim3(1), 0, comparingStream, device_data, 1);
+  // kernel will block, causing the first kernel to busy wait
+  hipLaunchKernelGGL(setting_kernel, dim3(1), dim3(1), 0, defaultStream, device_data, 1);
+
+  // test that nullStream blocks, hipStreamPerThread is non-blocking
+  if (defaultStream == nullptr) {
+    REQUIRE(streamBlocked(comparingStream, std::chrono::seconds(1)));
+
+    // kernel will set some memory, allowing the first kernel to complete
+    hipLaunchKernelGGL(setting_kernel, dim3(1), dim3(1), 0, settingStream, device_data, 1);
+  }
+
+  REQUIRE(!streamBlocked(comparingStream, std::chrono::seconds(1)));
+
+  HIP_CHECK(hipStreamDestroy(comparingStream));
+  HIP_CHECK(hipStreamDestroy(settingStream));
+  HIP_CHECK(hipFree(device_data));
+}
+
+TEST_CASE("Unit_hipStreamCreateWithFlags_ConcurrentNullStream") {
+  checkConcurrentSupported();
+
+  constexpr auto data_size = sizeof(int);
+  hipStream_t defaultStream = GENERATE(static_cast<hipStream_t>(nullptr), hipStreamPerThread);
+  INFO("Testing with default stream " << defaultStream);
+
+  int* device_data{};
+  HIP_CHECK(hipMalloc(&device_data, data_size));
+  HIP_CHECK(hipMemset(device_data, 0, data_size));
+
+  hipStream_t comparingStream{};
+  HIP_CHECK(hipStreamCreateWithFlags(&comparingStream, hipStreamNonBlocking));
+
+  // kernel will wait until the second kernel sets some memory
+  hipLaunchKernelGGL(comparing_kernel, dim3(1), dim3(1), 0, comparingStream, device_data, 1);
+  // kernel will set some memory, allowing the first kernel to complete
+  hipLaunchKernelGGL(setting_kernel, dim3(1), dim3(1), 0, defaultStream, device_data, 1);
```

Comment:
Assuming overlapping in such a way might not always work in case the streams are on the same HW queue and that queue is full from extra work.
We usually measure concurrency by measuring time of each kernels then in parallel and t1+t2 >= combined.

Reply:
I guess since it's just one thread, it might just be fine.